### PR TITLE
[RemoteMirror] Fix handling of generic MPEs

### PR DESCRIPTION
**Pull request by tbkka**
This exercises a number of cases that the previous logic got wrong, including cases where MPEs are laid out like SPEs, and cases where we use the SPE or MPE layout strategies for enums that have no non-empty payloads.  (These cases are superficially similar but differ in how they handle XIs.)

These new tests allowed me to correct a number logical flaws about how layouts are selected.  In particular, cases with generic payloads are always considered to be non-empty for purposes of selecting a layout strategy.  Only cases that are statically known to be empty are considered empty for layout purposes.

Resolves: ~~rdar://105709663~~, ~~rdar://107430963~~, rdar://106655244

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
TODO: I need to fill in the 32-bit tests before this can be merged
TODO: I've managed to construct a test that breaks this by having a generic case that is zero-sized but is not a simple `case a(T)`.

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Review by mikeash — APPROVED** [review]
Looks great, should be a nice improvement.

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Review by al45tair — APPROVED** [review]
LGTM.

**Comment by tbkka** [comment]
The one failing LLDB test on Linux seems to be related to an incomplete detection of indirect enum cases inside of LLDB.  I'm talking to the LLDB folks about how best to fix this.

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
I've also done a lot of local testing on Linux aarch64 and believe I've fixed everything except the one LLDB test that's asserting when trying to examine a simple Regex object.  As mentioned above, this seems to be related to an incomplete implementation of indirect enum cases within LLDB.

**Comment by tbkka** [comment]
CC: @adrian-prantl @kastiglione

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
Now that #65332 has landed, let's see if that makes any difference here.

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
Fixed a dumb bug in the SPE projection:  The number of cases stored in the _payload_ area depends on the size of the _payload_ area (not the size of the discriminator).

That seems to fix this particular failure locally.  Let's see if CI agrees.

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
All the CI failures are because the hosts were unable to check out sources from Github.  <sigh>

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
apple/llvm-project#6719

@swift-ci Please test

**Comment by tbkka** [comment]
Huh.  Failing to build LLDB?  apple/llvm-project#6719 has been merged, so let me try again.

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
LLDB fix took a while to get to where I needed it.   So try yet again....

**Comment by tbkka** [comment]
@swift-ci Please test
